### ISSUE TITLE
Logistic regression understanding

### ISSUE THREAD

**Issue by sebsfox** [nhs-bnssg-analytics/d_and_c#24]
- [ ] Are the warnings that are generated by the modelling important?
- [ ] Interaction terms
- [ ] Trend variable
- [ ] Offset
- [ ] using `glmnet` to penalise variables that have a detrimental effect on the modelling

**Comment by sebsfox** [comment]
# `glmnet`

When moving the logistic regression code over to `glmnet` and then apply the `tune()` function to selecting the `penalty` and `mixture` hyperparameters, I start to get errors when fitting the model. This looks to be because the outcome variable is a 2 column matrix `c(remainder, numerator)`. I have narrowed this down to an issue with `glmnet` rather than `tune()`.

## Testing `outcome = c(remainder, numerator)` compared with `outcome = value`

Using the same predictors, I have worked through fitting a model to predict both `val` and `c(remainder, numerator)` (both using `tidymodels` and raw `glm`). Fitting a model to predict `val` produces a model with different coefficients to a model predicting `c(remainder, numerator)` - though the different modelling approaches are consistent with each other (eg, tidymodels to predict val gives the same model as the glm model to predict val).

Investigating whether the differences above are to do with predicting a proportion versus rather than an odds ratio:

**Comment by sebsfox** [comment]
# warnings

I receive two warnings when running the glmnet model:

## glmnet.fit: algorithm did not converge

The argument λ is a coefficient in the glmnet equation. When tuning the `mixture` and `penalty` hyperparameters, this coefficient increases between two values, and model convergence is assessed. The `glmnet` function has an argument called `nlambda` which defaults to 100. This controls the maximum number of λ to assess model convergence over. If the number is too small, the steps between each λ will be bigger, and could result in difficult when trying to achieve convergence. This commit (2219f933c65247fb5d1cc4fd56d449b00f492710) increases the value for nlambda, which results in more models converging and the warning not appearing as often. I suppose the risk of this could be that models could be deemed to converge earlier than they otherwise would. When changing the value of `nlambda` from 100 to 600, the runtime increased, more models converged, but the performance didn't increase. Therefore it seemed there is little benefit of doing this.

**Comment by sebsfox** [comment]
# Not splitting data sequentially

Up until now, I have ordered the observations and separated the final year of data for testing. In theory, this was done with practical application in mind. I thought that if we could build a model on the years prior to the year that we want to use the model for, and the model performed well, then it would give users confidence in the model and it would encourage them to use it more. In practice, say I am testing data on 2022 based on data from 2020 and 2021, then we couldn't implement that 2020/2021 model for 2023 because we'd be missing a whole year of information. By mixing all of the years, this uses all of the latest information to build a model to make it as relevant for application as possible. It hugely improves performance and shows that the more information you feed into the model the better it performs:

![image](https://github.com/nhs-bnssg-analytics/d_and_c/assets/12894470/a4632f96-0a46-40ee-84ad-a4b8e2b9837c)

**Comment by sebsfox** [comment]
# Questioning the high performance of the model

The current performance of the models on the train, validation and test dataset is very good (best performing models have an R2 of 0.95, which is a little unbelievable). 

## Ensure I understand the effect of ridge and lasso regression

[This commit](dc56f2ae209d92d1752dab1847912a9083e33183) adds a script which demonstrates strong modelling performance, along with tweaking the `mixture` argument for the glmnet function to 0 (pure ridge regression) and 1 (pure lasso regression) to show that each of them perform as expected (ridge regression still retains every variable but lasso regression removes variables).

## Are there identical observations in train/val/test

If there are identical observations it makes it easy for the model to perform well. A way to test this is to perform "leave-p-out-cross-validation", where you can leave a group out in each fold of the cross validation process. [This commit](cabd6bda868f69dcd8409716c9782b567940a21b) implements this change, which has the effect of slightly lowering the performance of the model, but the performance is still very good:

![image](https://github.com/nhs-bnssg-analytics/d_and_c/assets/12894470/3112cccd-e285-487d-83cd-2488c8995b22)

## Understanding the variation in the target variable

The target variable has a strange distribution. The distribution has two humps, which are most likely COVID related, but need further investigation to make sure they are being calculated properly:

![image](https://github.com/nhs-bnssg-analytics/d_and_c/assets/12894470/ca969a15-6a29-4498-9db3-73b48abc7122)

## Error metric

I'm currently using R^2^ as the metric to tune the model on, but it gives an unusual shape when plotting observed versus predicted:

![image](https://github.com/nhs-bnssg-analytics/d_and_c/assets/12894470/5a2441c6-376b-4ba3-ac23-2f1a30b4859d)

Need to investigate other metrics.

## COVID variables

Is there something wrong with including the COVID variables. For many of the years their value is 0 for all ICSs (though this is centred and scaled). Will this have a strange effect on the modelling? At the moment, they don't come out particularly high when looking at variable importance.

## Dataset dimensions

Is there anything wrong with having a dataset that is wider than it is long?

**Comment by sebsfox** [comment]
# What variables are used in the final model?

[This article](https://parsnip.tidymodels.org/reference/glmnet-details.html) is really useful to understand how `glmnet` is applied through {parsnip}. There are two values that are tuned in the hyperparameter tuning process; `penalty` (aka lambda - the coefficient applied to penalise the error terms with) and `mixture` (the extent of either lasso or ridge regression).

The hyperparameter tuning process selects the two parameters above that produce the best performance on the validation dataset. However, despite `penalty` (lambda) having an optimal value, when fitting the best model to the final training dataset, the `glmnet()` process still follows a "lambda path" of nlambda values, whose range will contain the optimal `penalty` (lambda). The parameters used in the final fit will be the ones that still remain when the lambda path is nearest the optimal `penalty`.

**Comment by sebsfox** [comment]
When the convergence warning occurs, the warnings are for some of the 30 combinations of hyperparameters for some of the folds when cross validation is implemented. These combinations do not end up being selected by the optimisation process as the best set of hyperparameters

**Comment by sebsfox** [comment]
To add, interaction terms and trend terms haven't been included and could be added to the backlog for future improvement
